server/src/main/kotlin/com/tilbox/api/security/EmailPasswordAuthenticationFilter.kt

```diff
@@ -0,0 +1,45 @@
+package com.tilbox.api.security
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.core.log.LogMessage
+import org.springframework.http.HttpMethod
+import org.springframework.security.authentication.AuthenticationManager
+import org.springframework.security.authentication.AuthenticationServiceException
+import org.springframework.security.authentication.event.InteractiveAuthenticationSuccessEvent
+import org.springframework.security.core.Authentication
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
+import javax.servlet.FilterChain
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+class EmailPasswordAuthenticationFilter(
+    authenticationManager: AuthenticationManager,
+    private val objectMapper: ObjectMapper
+) : UsernamePasswordAuthenticationFilter(authenticationManager) {
+    override fun attemptAuthentication(request: HttpServletRequest, response: HttpServletResponse?): Authentication? {
+        if (!request.method.equals(HttpMethod.POST.name)) {
+            throw AuthenticationServiceException("Authentication method not supported: " + request.method)
+        }
+        val loginRequest = objectMapper.readValue(request.inputStream, LoginRequest::class.java)
+        setDetails(request, loginRequest.toAuthRequest())
+        return authenticationManager.authenticate(loginRequest.toAuthRequest())
+    }
+
+    override fun successfulAuthentication(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        chain: FilterChain,
+        authResult: Authentication
+    ) {
+        SecurityContextHolder.getContext().authentication = authResult
+        if (logger.isDebugEnabled) {
+            logger.debug(LogMessage.format("Set SecurityContextHolder to %s", authResult))
+        }
+        rememberMeServices.loginSuccess(request, response, authResult)
+        if (eventPublisher != null) {
+            eventPublisher.publishEvent(InteractiveAuthenticationSuccessEvent(authResult, this::class.java))
+        }
```

Comment:
이 메서드의 역할이 궁금합니다! (핑프 지송 ..ㅎㅎ )

Reply:
스프링 시큐리티에서 제공하는 기본 form login 인증방식`UsernamePasswordAuthenticationFilter` 구현체를 일부 변형하여 요청 body를 파싱 후 인증하는 필터를 만드는 과정에서 발생한 코드에 포함되었네요. `UsernamePasswordAuthenticationFilter`을 override 하는 과정에서 커스터마이징 제약이 걸리는 부분이 있어 부득이하게 포함된 것 같습니다.

저도 궁금해져서 `InteractiveAuthenticationSuccessEvent`에 대해서 간단히 찾아봤는데, 사용 목적이 자세하게 나오는 곳은 없네요. 시큐리티 공식 문서에도 별도의 언급은 없어서 한번 클래스 헤더에 있는 주석을 가져와봤습니다.

```markdown
Indicates an interactive authentication was successful.
The ApplicationEvent's source will be the Authentication object.
This does not extend from AuthenticationSuccessEvent to avoid duplicate AuthenticationSuccessEvents being sent to any listeners.
```

대화식 인증이 성공했음을 알리는 용도로 사용되는 이벤트 인 것 같습니다. 인증 성공 후 생성한 인증 객체와 인증을 수행한 클래스를 이벤트 내용으로 포함합니다. 그 이상의 설명은 없는것으로 보아, 시큐리티 사용자가 로그인 후에 커스텀한 동작을 수행하도록 이벤트를 제공해 주는 것 같습니다.

![image](https://user-images.githubusercontent.com/20358042/144734280-8259b8a4-9b70-499a-b527-c54d3b3425fe.png)

공식 문서에 포함된 이미지인데, 인증 성공 후에 `ApplicationEventPublisher`가 발행하는 이벤트라고 하네요. 어디서 사용할 수 있는지 고민해보니, 쓸만한 곳이 있을 것 같기는 해요. 로그인시 앱에 로그인 알림 푸시를 보내거나, 로그인 정보를 어딘가에 기록하기 위해 사용할 수 있을 것 같습니다.